core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -249,6 +247,10 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime, threadNamePr
         registerStats()
 
         brokerState.newState(RunningAsBroker)
+
+        /* start kafka controller after setting state RunningAsBroker as it subsequently become controller */
+        kafkaController.startup()
```

Comment:
One issue with this is that there's a short window where `KafkaController.shutdownBroker` can be called from `KafkaApis` before `startup()` is called. Furthermore, since `KafkaController` is available as a field in `KafkaApis`, more calls could be added in the future and it could lead to surprising behaviour.

One thing that could potentially be done is to pass a function to `KafkaApis` that returns the controller only if it's already initialised (and maybe throws an exception otherwise).

But it all seems a bit too complicated. I wonder if `RunningAsController` is actually a useful state. It seems better to have a separate boolean indicating whether the broker is also a controller (`KafkaController.isActive`) instead of encoding it via broker states.

Thoughts @theduderog and @junrao?